File: problems/binary_trees/flatten_binary_tree/js/flatten_binary_tree.js
export const flattenBinaryTree = (() => {
  const flattenInOrder = (node, state) => {
    if (!node) return;

    // traverse left subtree
    flattenInOrder(node.left, state);

    // store the head of the linked list
    if (!state.head) state.head = node;
    // link the previous node with the current node
    if (state.prev) {
      state.prev.right = node;
      node.left = state.prev;
    }
    // update the previous node
    state.prev = node;

    // traverse right subtree
    flattenInOrder(node.right, state);
  };

  return (root) => {
    const state = { prev: null, head: null };
    flattenInOrder(root, state);

    return state.head;
  };
})();

File: problems/binary_trees/flatten_binary_tree/python/flatten_binary_tree.py
from problems.binary_trees.assets.BinaryTree import BinaryTree


def flattenBinaryTree(root: BinaryTree) -> BinaryTree:

    def flattenInOrder(node: BinaryTree, state: dict) -> None:
        if not node:
            return

        flattenInOrder(node.left, state)

        if not state["head"]:
            state["head"] = node

        if state["prev"]:
            state["prev"].right = node
            node.left = state["prev"]

        state["prev"] = node

        flattenInOrder(node.right, state)

    state: dict = {"prev": None, "head": None}
    flattenInOrder(root, state)

    return state["head"]
